Le Log

underdoeg opened the pull request:
logging module

underdoeg committed:
added le_log

underdoeg commented on modules/le_log/le_log.cpp:
```diff
@@ -0,0 +1,94 @@
+	std::atomic_int log_level = 1;
+};
+
+static le_log_module_o *le_log_module_default() {
```
It might make sense to have a logging initialization where a default context is created. but not sure where to place this startup method.

tgfrerer replied:
i saw you placed it in the api registration method, which was going to be the place i was going to suggest to put it :)

underdoeg replied:
yeah, took me a second to realize that this method gets called each time with the same api object.

tgfrerer commented on modules/le_log/le_log.cpp:
```diff
@@ -0,0 +1,94 @@
+}
+
+static le_log_module_o *le_log_get_module( const char *name ) {
+	static std::unordered_map<std::string, le_log_module_o *> modules;
```
i have a feeling that this will leak on hot-reload, as `modules` will be re-initialised, and all previously held elements in modules will get leaked, because the c++ delete operation will only delete the pointers to `le_log_module_o`, but not the objects themselves.

in any case, we don't want the objects to drop until the application is unloaded. the safest way to keep them around is to store a pointer to the collection with the api.

underdoeg replied:
Yes because these will be reinitialized, I was wondering about persistent static data. or maybe initialize one global pointer on startup. but since all other modules will probably rely on le log, this initialization method will always have to be called

tgfrerer replied:
Re: persistent static data: [here](https://github.com/tgfrerer/island/blob/a9e7d2391b9c6e79d4a8a03dd2f1833b3776cdc1/modules/le_renderer/le_renderer.cpp#L730-L732
) is how I implemented that pattern for le_renderer's texture_handle_store.

tgfrerer commented on modules/le_log/le_log.cpp:
```diff
@@ -0,0 +1,94 @@
+};
+
+static le_log_module_o *le_log_module_default() {
+	static auto module = new le_log_module_o();
```
this won't be the same after hot-reload, as hot-reloading means invalidating and re-initializing all static memory.

underdoeg replied:
I'll try a similar approach to the one use for the textures in le_renderer

underdoeg reviewed (commented):
--

underdoeg committed:
le_log add persistent context

underdoeg commented:
the new commit might work, basically a persistent logging context that gets created or reused on module registration.

tgfrerer commented on modules/le_log/le_log.cpp:
```diff
@@ -11,23 +11,28 @@ struct le_log_module_o {
+	return &ctx->module_default;
 }
 
 static le_log_module_o *le_log_get_module( const char *name ) {
```
What happens if this gets called with `le_log_get_module("")` (empty string)? I think it could be good to then return the default le_log module

underdoeg replied:
I suggest that we crash the whole system and display a message "trying to outsmart me, arr" :)
no, makes sense. I have added a check

tgfrerer commented on modules/le_log/le_log.cpp:
```diff
@@ -91,4 +96,10 @@ LE_MODULE_REGISTER_IMPL( le_log, api ) {
 	le_api->error      = le_log_implementation<le_log::Level::ERROR>;
 	le_api->get_module = le_log_get_module;
 	le_api->set_level  = le_log_set_level;
+
```
looks good!

underdoeg committed:
le_log check for empty or invalid module name

tgfrerer commented on modules/le_log/le_log.h:
```diff
@@ -0,0 +1,103 @@
+// clang-format off
+struct le_log_api {
+
+    void ( *set_level  )(le_log_module_o *module, le_log::Level level);
```
i feel it would be helpful to have a logger interface here, as we do with other modules.

there might be some le_log related functionality which is not related to individual `le_log_module_o` objects, (as in, for example a possibly future feature to **globally** silence any log messages below a certain log level), and it will then be easier to keep things organised if we keep all le_log_module_o* related functions in their own interface. 

We can think of interfaces as the class namespace for functions which can be applied on `le_log_module_o` objects. placing this inside of a struct helps us in another way: it provides a namespace to clients of the c api. 

My hope here is that it will help with consistency and discoverability, as methods on objects from `le_module`s are usually called via their object's interface, and not their parent api interface.

I suggest naming the interface `le_log_module_interface_t` and grouping L21-31 under this interface...

underdoeg replied:
can do

underdoeg committed:
add le_log module interface

tgfrerer committed:
[le_log] move enum into api namespace

we're moving enum into api so that it is namespaced even for c consumers
of the api.

For the c++ facade, we import the enum into the le_log namespace.

tgfrerer committed:
[le_log] whitespace reorg

move creator/getter to be the first method visible in the header file,
group similar methods closer together.

tgfrerer committed:
[le_log] add namespace alias for le_log_i

tgfrerer committed:
[le_log] rename namespace le_log -> le::Log

could either be LeLog or le::Log, I prefer le::Log

we're trying to use CamelCase for Namespaces, as they are c++ ish, and
snake_case for things which are more c-ish.

tgfrerer committed:
[le_log] add missing static specifier

methods like these which are implemented in headers need to be declared
static, otherwise we end up with linker errors when compiling for
Release (where we create a static binary).

tgfrerer commented:
I've tested on my local system and things are working beautifully! 

I have a question re: keeping a global static log context - can you let me know a bit more about the reasons and intuitions that informed your design decision there? 

I wonder what speaks against having loggers be regular objects (with regular `create()` `destroy()` methods), which are not owned globally by their api, but by whoever creates them - which is the default throughout the api. 

The default approach would allow us to clear up and destroy any loggers after we're done using them on app teardown for example. The current approach means that we must keep loggers alive for as long as the api is alive (forever) which means we inevitably leak each logger that we created, on app teardown. 

Are there any major benefits from having loggers be globally owned by their API?

underdoeg commented:
We could already destroy individual logger modules, just not the entire context atm? I think it makes sense to keep the default logger forever. It is not something that has to be destroyed until the very end (where we could use a deconstructor to cleanup).

The way I like to use a logger is having the ability to dynamically create any output channels and verbosity for all my modules on the fly. Usually by having something like this in the application code.

```
log::get("render").set_level(WARN);
log::get("asset_loader").set_level(INFO);

log::get("sensor").set_output("some_file.log");
...
```
These settings have to be persistent on reloads.

Alternatively each module could expose it's logging module in the h file.

```
le_log::set_level(le_render::log, DEBUG);
```

But that would include more manual work which I think can be avoided by addressing loggers by string keys.

so if I see some module spamming my terminal, I can just use the name and silence it from anywhere.

I think of it as a global persistent database of all loggers.

[EDIT] I am generally with you. global static is bad and needs to be avoided. this might be the only use case, I can think of, where it makes total sense to me.

tgfrerer commented:
At the moment logger modules live inside the logger context, and I fear destroying them could leave the logger context with a dangling reference, unless we managed to destroy them via the logger context. 

In this case, we would have to find a way to enforce that users of the logger can't store direct pointers to `le_log_module_o`, since there would be no way of telling whether such a pointer will be valid or not (any pointers to `le_log_module_o` would have to be seen as weak pointers)

If we address the loggers via string constants, there is a chance that we address the same logger from two different contexts and these two contexts might fight over the logger - what happens, for example if i set the `log_level` for logger `my_logger` to `debug` when I'm inside le_renderer, but then, inside le_backend, i'm setting the log_level to `info` - this sounds like the state of the logger from inside backend will then be the log level that's kept until somebody else changes it back again. I fear that this could cause quite a bit of confusion, because it's not evident where the last change to the logger object happened. 

And what if two threads share a logger? and they fight for log levels? wouldn't it be easier if each thread had their own logger? 

Perhaps I'm overly pessimistic, but with public apis, if there is a way to design possible bugs out of the system, I feel it could be worth it :)

underdoeg commented:
I am not entirely sure I follow by how you want to ban le_backend from setting the log level of le_renderer while still allowing any user to disable / enable logging for certain modules? 
IMO a module itself should never change its log level (except maybe a default level on startup) or the level of any other module. leave this entirely up to the user. The same applies to different threads. They just log to a certain level but will never set it.

How would you implement the code for the logger of le_render to avoid your scenario, while still maintaining the user selected log level on library reload?

It might make sense to add an optional argument to get_module with the default log level. which will be used if the logger is newly created but otherwise ignored.

I agree on the dangling logger pointers that don't really have a clear owner (except for the loggign context). Then on the other hand you shouldn't create too many loggers and the objects themselves.

le_log_module_o only exists to speed up logging performance and simplify multithreading. It would be way easier to have a map will all log levels.

tgfrerer commented:
I think the tricky bit is that if ownership is held globally inside the logger api, all accesses to loggers will be global and have the potential stepping onto each other's toes - imagine a logger/channel called "fancy_debug" which is used inside `le_backend` and `le_renderer` - they would potentially fight over internal state of that channel.

How about if both, the renderer and the backend, separately create logger objects, which they then own internally?

We could still filter for strings if we changed the global filter settings in the logger output module - these filter settings could be stored with the logger api/context. 

The user-selected log-level is maintained, because a logger object will be created and live on the heap, as will its owner (le_renderer for example). Anything that we allocate on the heap stays alive when we hot-reload.

tgfrerer commented:
Ahh, okay, i see - having logger channels globally allows you to set log levels explicitly when debugging... there is a case for that, i agree. let's see how it goes. 

i'll rename and update the header file as we discussed in our chat, will send an update shortly...

tgfrerer committed:
[le_log] rename le_log_module -> le_log_channel

tgfrerer committed:
[le_log] add cpp facades

Add facade for usinga specific logger channel via le::Log object, and
le::log_[warn|info|error|debug] methods for logging to the default
channel.

tgfrerer committed:
[le_log] clarify log level check

tgfrerer committed:
[le_log] flush file stream after log

not flushing makes for faster rendering, but we don't want to miss
messages... we might change this behaviour if we find it slows things
down too much.

tgfrerer committed:
[le_log] make enum type same type as internal atomic type

tgfrerer committed:
[le_log] fix error in facade

was not forwarding channel

tgfrerer committed:
[le_log] add test app for le_log

add example app for le_log, which shows how le_log works, and which
doubles as a unit test for le_log

tgfrerer commented:
okay, this is almost ready... only it breaks the windows integration test. i need to switch over to windows in order to see what's going on.

tgfrerer committed:
[le_log] cleanup cmake file

modernizing cmake file in the hope that this will fix our windows
compilation errors.

tgfrerer committed:
[scripts] update template file for modules

tgfrerer commented:
Allright, integration tests have passed - looks ready to merge, unless there are any last-minute reservations?

underdoeg commented:
looking good, go for it

underdoeg commented:
next step is to integrate the log into the existing modules

tgfrerer commented:
Merged. Thank you!!

underdoeg commented:
nice

tgfrerer commented:
> next step is to integrate the log into the existing modules

Yes, that would be nice. This might be a bit tricky with `le_core` as this is loaded before the `le_log` module, but for the rest it should be fairly straightforward as long as the log-level is self-evident.
